### ISSUE TITLE
wxSTC doesn't respond to dark mode

### ISSUE THREAD

**Issue by wxtrac** [wxWidgets/wxWidgets#19107]
*Issue migrated from trac ticket # 19107*

**component:** wxStyledText | **priority:** normal

#### 2021-03-19 17:23:15: @imciner2 created the issue

___
The wxSTC control doesn't respond to dark mode like the other controls do (e.g. native text controls) - it simply stays with a white background while everything else changes.

This is evident in the stc sample application when running under a dark theme on GTK.

**Comment by vadz** [comment]
I can set the colours appropriate for the dark mode by default, but the problem is that, AFAICS, all the existing code using wxSTC (including our own sample) sets its own colours for it, so they would override any default colours anyhow.

But @pkulchenko is correct that there is no straightforward way to make "light colours" work in dark mode. I've tried to apply the same transformation as I already used elsewhere for the colours, i.e. map HSV value `v` to `sqrt(1-v^2)`, but this makes the default dark blue used for C++ keywords in the stc sample totally unreadable on black background and this was just the first example I tried. I've also tried inverting foreground and background and this was more readable but just plain ugly...

So I think #23097 is the best we can do: it does set the default colours correctly, but still requires the application to be modified in the same (or similar) way as the stc sample was.

Please let me know if anybody sees anything better to do here.

**Comment by NewPagodi** [comment]
I think it might be best to leave it to the user to respond to dark mode as they see fit.  The reason is that there is a built in `StyleResetDefault` method that will set the control to use colors defined in from the Scintilla library.  If we initially set some colors in response to dark mode, a user might expect the those colors to be restored when the StyleResetDefault function is called, but the ones from Scintilla will be used instead which might be confusing.

Alternately, maybe an extra function called `StyleResetForDarkMode`could be added to try to set some initial colors that might be appropriate for a dark theme.

**Comment by eranif** [comment]
I would disagree here.
I often use `wxSTC` as a replacement for `wxTextCtrl` without any special style settings. This means that if the default is not appropriate for the current app theme, it just looks misplaced.

**Comment by vadz** [comment]
I thought about making this opt-in, but the existence of `StyleClearAll()` (and its prevalence in the existing code) has finally convinced me _not_ to do it: AFAICS most of the existing code setting its own colours does something similar to
```cpp
    StyleSetForeground(wxSTC_STYLE_DEFAULT, *wxBLACK);
    StyleSetBackground(wxSTC_STYLE_DEFAULT, *wxWHITE);
    StyleClearAll();
```
and this will still continue to work in exactly the same way because this will override the colours we set by default.

If you don't set any special colours, however, the control will now use more appropriate colours by default (and not only under MSW, but under the other systems too).

**Comment by vadz** [comment]
I still don't think there is anything more that we could do here, so this will be closed by merging #23097.
